File: oolab/src/main/java/agh/ics/oop/model/RandomPositionGenerator.java
package agh.ics.oop.model;

import java.util.ArrayList;
import java.util.NoSuchElementException;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

class RandomPositionGenerator implements Iterable<Vector2d> {
    final int maxWidth;
    final int maxHeight;
    final int grassCount;

    public RandomPositionGenerator(int maxWidth, int maxHeight, int grassCount) {
        this.maxWidth = maxWidth;
        this.maxHeight = maxHeight;
        this.grassCount = grassCount;
    }

    @Override
    public Iterator<Vector2d> iterator() {
        return new Iterator<>() {
            private final List<Vector2d> positions = generateRandomPositions();

            private List<Vector2d> generateRandomPositions() {
                List<Vector2d> allPositions = new ArrayList<>();
                for (int x = 0; x < maxWidth; x++) {
                    for (int y = 0; y < maxHeight; y++) {
                        allPositions.add(new Vector2d(x, y));
                    }
                }

                Collections.shuffle(allPositions);
                return allPositions.subList(0, grassCount);
            }

            private int currentIndex = 0;

            @Override
            public boolean hasNext() {
                return currentIndex < grassCount;
            }

            @Override
            public Vector2d next() {
                if (!hasNext()) {
                    throw new NoSuchElementException("No more positions to generate.");
                }
                return positions.get(currentIndex++);
            }
        };
    }
}

File: oolab/kotlin/src/main/kotlin/model/utils/RandomPositionGenerator.kt
package model.utils

import model.Vector2d

class RandomPositionGenerator(private val maxWidth: Int, private val maxHeight: Int, private val grassCount: Int) :
    Iterable<Vector2d> {

    private val positions: List<Vector2d> by lazy { generateRandomPositions() }
    private var currentIndex = 0

    init {
        require(grassCount <= maxWidth * maxHeight) { "grassCount must be less than or equal to maxWidth * maxHeight" }
    }

    private fun generateRandomPositions(): List<Vector2d> {
        val allPositions = mutableListOf<Vector2d>()
        for (x in 0 until maxWidth) {
            for (y in 0 until maxHeight) {
                allPositions.add(Vector2d(x, y))
            }
        }

        allPositions.shuffle()
        return allPositions.subList(0, grassCount)
    }

    override fun iterator(): Iterator<Vector2d> {
        return object : Iterator<Vector2d> {
            override fun hasNext(): Boolean {
                return currentIndex < grassCount
            }

            override fun next(): Vector2d {
                if (!hasNext()) {
                    throw NoSuchElementException("No more positions to generate.")
                }
                return positions[currentIndex++]
            }
        }
    }
}
